lib/nbd.c

```diff
@@ -0,0 +1,792 @@
+/***************************************************************************
+ *                      _   _ ____  _
+ *  Project         ___| | | |  _ \| |
+ *                 / __| | | | |_) | |
+ *                | (__| |_| |  _ <| |___
+ *                 \___|\___/|_| \_\_____|
+ *
+ * Copyright (C) 2018, Richard W.M. Jones, <rjones@redhat.com>
+ * Copyright (C) 2018, Red Hat Inc.
+ *
+ * This software is licensed as described in the file COPYING, which
+ * you should have received as part of this distribution. The terms
+ * are also available at https://curl.haxx.se/docs/copyright.html.
+ *
+ * You may opt to use, copy, modify, merge, publish, distribute and/or sell
+ * copies of the Software, and permit persons to whom the Software is
+ * furnished to do so, under the terms of the COPYING file.
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+ * KIND, either express or implied.
+ *
+ ***************************************************************************/
+
+#include "curl_setup.h"
+
+#ifndef CURL_DISABLE_NBD
+
+/* Curl support for the NBD protocol.
+ *
+ * Only simple uploads and downloads of the whole image are supported.
+ * Only servers using the "fixed newstyle" protocol are supported.
+ *
+ * Some other missing features:
+ *
+ * - Connect over Unix domain socket.
+ *
+ * - TLS authentication and encryption.
+ *
+ * - Detect if upload block is all zeroes and use NBD_CMD_WRITE_ZEROS.
+ *
+ * - Range requests (upload/download part of the image).
+ *
+ * - Support for: trim, write zeroes, flush.
+ *
+ * - Setting FUA flag in requests.
+ *
+ * - Support for server flags: readonly, rotational.
+ *
+ * - List export names.
+ */
+
+#ifdef HAVE_NETINET_IN_H
+#include <netinet/in.h>
+#endif
+#ifdef HAVE_ARPA_INET_H
+#include <arpa/inet.h>
+#endif
+#ifdef HAVE_UTSNAME_H
+#include <sys/utsname.h>
+#endif
+#ifdef HAVE_NETDB_H
+#include <netdb.h>
+#endif
+#ifdef __VMS
+#include <in.h>
+#include <inet.h>
+#endif
+
+#if (defined(NETWARE) && defined(__NOVELL_LIBC__))
+#undef in_addr_t
+#define in_addr_t unsigned long
+#endif
+
+#include <curl/curl.h>
+#include "urldata.h"
+#include "connect.h"
+#include "progress.h"
+#include "sendf.h"
+#include "multiif.h"
+#include "transfer.h"
+#include "escape.h"
+#include "nbd.h"
+/* The last #include file should be: */
+#include "memdebug.h"
```

Comment:
I recommend stealing the three last #include files from for example lib/url.c:
~~~c
/* The last 3 #include files should be in this order */
#include "curl_printf.h"
#include "curl_memory.h"
#include "memdebug.h"